File: Packages/RideSharingFeature/Sources/RideSharingFeature/RideMapGeometry.swift
import SwiftUI

enum RideMapGeometry {
    static let viewportSize = MapSize(width: 402, height: 740)

    static let origin = MapPoint(x: 196, y: 470)
    static let destination = MapPoint(x: 250, y: 165)
    static let transfer = MapPoint(x: 150, y: 320)

    static let waterBlock = MapBlock(
        center: MapPoint(x: 70, y: 670),
        size: MapSize(width: 220, height: 260),
        cornerRadius: 0,
        opacity: 0.9,
        rotationDegrees: -8
    )
    static let parkBlock = MapBlock(
        center: MapPoint(x: 370, y: 130),
        size: MapSize(width: 240, height: 180),
        cornerRadius: 10,
        opacity: 0.55,
        rotationDegrees: 0
    )
    static let districtBlock = MapBlock(
        center: MapPoint(x: 55, y: 195),
        size: MapSize(width: 150, height: 150),
        cornerRadius: 8,
        opacity: 0.50,
        rotationDegrees: 0
    )

    static let majorStreetWidth = 9.0
    static let majorStreetOpacity = 0.95
    static let minorStreetWidth = 4.0
    static let minorStreetOpacity = 0.60
    static let routeStrokeWidth = 7.0
    static let routeShadowRadius = 3.0
    static let routeShadowYOffset = 2.0
    static let routeShadowOpacity = 0.35
    static let transferRadius = 5.0
    static let transferStrokeWidth = 3.0
    static let currentPulseSize = 66.0
    static let currentPulseScaleStart = 0.33
    static let currentPulseOpacityStart = 0.65
    static let currentPulseDuration = 2.0
    static let currentDotSize = 22.0
    static let currentDotStrokeWidth = 3.0
    static let radarPulseSize = 126.0
    static let radarPulseScaleStart = 0.11
    static let radarPulseOpacityStart = 0.50
    static let radarPulseDuration = 1.8
    static let radarDotSize = 14.0
    static let radarDotStrokeWidth = 3.0

    static let majorStreets = [
        MapStreet(id: "m1", start: MapPoint(x: -20, y: 250), end: MapPoint(x: 430, y: 225)),
        MapStreet(id: "m2", start: MapPoint(x: -20, y: 370), end: MapPoint(x: 430, y: 350)),
        MapStreet(id: "m3", start: MapPoint(x: -20, y: 500), end: MapPoint(x: 430, y: 520)),
        MapStreet(id: "m4", start: MapPoint(x: -20, y: 630), end: MapPoint(x: 430, y: 650)),
        MapStreet(id: "m5", start: MapPoint(x: 70, y: -20), end: MapPoint(x: 120, y: 780)),
        MapStreet(id: "m6", start: MapPoint(x: 210, y: -20), end: MapPoint(x: 240, y: 780)),
        MapStreet(id: "m7", start: MapPoint(x: 330, y: -20), end: MapPoint(x: 360, y: 780))
    ]

    static let minorStreets = [
        MapStreet(id: "n1", start: MapPoint(x: -20, y: 180), end: MapPoint(x: 430, y: 165)),
        MapStreet(id: "n2", start: MapPoint(x: -20, y: 430), end: MapPoint(x: 430, y: 445)),
        MapStreet(id: "n3", start: MapPoint(x: 140, y: -20), end: MapPoint(x: 170, y: 780)),
        MapStreet(id: "n4", start: MapPoint(x: 280, y: -20), end: MapPoint(x: 305, y: 780))
    ]

    static let singleLegRouteControls = RouteControls(
        firstControl: MapPoint(x: 170, y: 400),
        secondControl: MapPoint(x: 300, y: 330),
        end: destination
    )

    static let multiLegRouteControls = [
        RouteControls(
            firstControl: MapPoint(x: 150, y: 430),
            secondControl: MapPoint(x: 120, y: 380),
            end: transfer
        ),
        RouteControls(
            firstControl: MapPoint(x: 175, y: 285),
            secondControl: MapPoint(x: 230, y: 230),
            end: destination
        )
    ]

    static func carPoints(isMultiLeg: Bool) -> [MapPoint] {
        if isMultiLeg {
            return [origin, MapPoint(x: 150, y: 400), transfer, MapPoint(x: 205, y: 250), destination]
        }
        return [origin, MapPoint(x: 215, y: 390), MapPoint(x: 285, y: 300), destination]
    }
}

struct MapPoint {
    let x: Double
    let y: Double
}

struct MapSize {
    let width: Double
    let height: Double
}

struct MapBlock {
    let center: MapPoint
    let size: MapSize
    let cornerRadius: Double
    let opacity: Double
    let rotationDegrees: Double
}

struct RouteControls {
    let firstControl: MapPoint
    let secondControl: MapPoint
    let end: MapPoint
}

struct MapStreet {
    let id: String
    let start: MapPoint
    let end: MapPoint

    func path(in viewport: MapSvgViewport) -> Path {
        var path = Path()
        path.move(to: viewport.point(start))
        path.addLine(to: viewport.point(end))
        return path
    }
}

struct MapSvgViewport {
    let scale: Double
    let offsetX: Double
    let offsetY: Double

    init(size: CGSize) {
        let viewport = RideMapGeometry.viewportSize
        scale = max(Double(size.width) / viewport.width, Double(size.height) / viewport.height)
        offsetX = (Double(size.width) - viewport.width * scale) / 2
        offsetY = (Double(size.height) - viewport.height * scale) / 2
    }

    func point(_ point: MapPoint) -> CGPoint {
        CGPoint(x: offsetX + point.x * scale, y: offsetY + point.y * scale)
    }

    func length(_ value: Double) -> CGFloat {
        CGFloat(value * scale)
    }
}

File: android/app/src/main/java/com/liive/ride/ui/RideMapGeometry.kt
package com.liive.ride.ui

import androidx.compose.ui.geometry.Offset

internal object RideMapGeometry {
    const val MapWidth = 402f
    const val MapHeight = 740f

    val Origin = MapPoint(196f, 470f)
    val Destination = MapPoint(250f, 165f)
    val Transfer = MapPoint(150f, 320f)

    val WaterBlock = MapBlock(
        topLeft = MapPoint(-40f, 540f),
        pivot = MapPoint(70f, 670f),
        size = MapSize(220f, 260f),
        cornerRadius = 0f,
        alpha = 0.9f,
        rotationDegrees = -8f
    )
    val ParkBlock = MapBlock(
        topLeft = MapPoint(250f, 40f),
        pivot = MapPoint(370f, 130f),
        size = MapSize(240f, 180f),
        cornerRadius = 10f,
        alpha = 0.55f,
        rotationDegrees = 0f
    )
    val DistrictBlock = MapBlock(
        topLeft = MapPoint(-20f, 120f),
        pivot = MapPoint(55f, 195f),
        size = MapSize(150f, 150f),
        cornerRadius = 8f,
        alpha = 0.50f,
        rotationDegrees = 0f
    )

    const val MajorStreetWidth = 9f
    const val MajorStreetAlpha = 0.95f
    const val MinorStreetWidth = 4f
    const val MinorStreetAlpha = 0.60f
    const val RouteStrokeWidth = 7f
    const val RouteShadowBlur = 3f
    const val RouteShadowYOffset = 2f
    const val RouteShadowAlpha = 0.35f
    const val TransferRadius = 5f
    const val TransferStrokeWidth = 3f
    const val CurrentPulseSize = 66
    const val CurrentPulseScaleStart = 0.35f
    const val CurrentPulseDurationMs = 2_000
    const val CurrentDotRadius = 11
    const val CurrentDotStrokeWidth = 3
    const val RadarPulseSize = 126
    const val RadarPulseScaleStart = 0.11f
    const val RadarPulseDurationMs = 1_800
    const val RadarDotRadius = 7
    const val RadarDotStrokeWidth = 3

    val MajorStreets = listOf(
        MapLine(-20f, 250f, 430f, 225f),
        MapLine(-20f, 370f, 430f, 350f),
        MapLine(-20f, 500f, 430f, 520f),
        MapLine(-20f, 630f, 430f, 650f),
        MapLine(70f, -20f, 120f, 780f),
        MapLine(210f, -20f, 240f, 780f),
        MapLine(330f, -20f, 360f, 780f),
    )

    val MinorStreets = listOf(
        MapLine(-20f, 180f, 430f, 165f),
        MapLine(-20f, 430f, 430f, 445f),
        MapLine(140f, -20f, 170f, 780f),
        MapLine(280f, -20f, 305f, 780f),
    )

    val SingleLegControls = RouteControls(
        firstControl = MapPoint(170f, 400f),
        secondControl = MapPoint(300f, 330f),
        end = Destination,
    )

    val MultiLegControls = listOf(
        RouteControls(
            firstControl = MapPoint(150f, 430f),
            secondControl = MapPoint(120f, 380f),
            end = Transfer,
        ),
        RouteControls(
            firstControl = MapPoint(175f, 285f),
            secondControl = MapPoint(230f, 230f),
            end = Destination,
        ),
    )

    fun carPoints(multiLeg: Boolean): List<MapPoint> =
        if (multiLeg) {
            listOf(Origin, MapPoint(150f, 400f), Transfer, MapPoint(205f, 250f), Destination)
        } else {
            listOf(Origin, MapPoint(215f, 390f), MapPoint(285f, 300f), Destination)
        }
}

internal data class MapPoint(val x: Float, val y: Float)
internal data class MapSize(val width: Float, val height: Float)

internal data class MapBlock(
    val topLeft: MapPoint,
    val pivot: MapPoint,
    val size: MapSize,
    val cornerRadius: Float,
    val alpha: Float,
    val rotationDegrees: Float,
)

internal data class RouteControls(
    val firstControl: MapPoint,
    val secondControl: MapPoint,
    val end: MapPoint,
)

internal data class MapLine(val x1: Float, val y1: Float, val x2: Float, val y2: Float)

internal data class MapSvgViewport(val width: Float, val height: Float) {
    private val scale = maxOf(width / RideMapGeometry.MapWidth, height / RideMapGeometry.MapHeight)
    private val offsetX = (width - RideMapGeometry.MapWidth * scale) / 2f
    private val offsetY = (height - RideMapGeometry.MapHeight * scale) / 2f

    fun point(point: MapPoint) = point(point.x, point.y)

    fun point(x: Float, y: Float) = Offset(offsetX + x * scale, offsetY + y * scale)

    fun length(value: Float) = value * scale
}
